crypto/x509/x509_cmp.c

```diff
@@ -257,9 +257,9 @@ unsigned long X509_NAME_hash(const X509_NAME *x)
 
     /* Make sure X509_NAME structure contains valid cached encoding */
     i2d_X509_NAME(x, NULL);
-    if (!EVP_Digest(x->canon_enc, x->canon_enclen, md, NULL, EVP_sha1(),
-                    NULL))
-        return 0;
+    if (!EVP_Digest(x->canon_enc, x->canon_enclen, md, NULL, EVP_sha1(), NULL))
+        /* TODO(3.0) This fails if SHA1 is not supported, e.g., in test_ssl_old */
+        return (unsigned long)0x100000000L; /* may collapse to 0 */
```

Comment:
I still think this is not a proper solution as except for LP64 architectures this will return 0 as error. Or to be more precise, it is not a problem here but all the callers checking for 0 as failure value. If you really want to check for the error condition the only proper solution IMO is to add another API call which would have a proper error reporting and deprecate this one.

Reply:
So I've spent the effort to add a new function `X509_NAME_hash_ex()` in order to be able to check 
if it failed due to unsupported SHA1 (or any other error reason), 
to deprecate `X509_NAME_hash()`,
and to document `X509_NAME_hash_ex()`, `X509_NAME_hash()` as well as the strongly related functions `X509_subject_name_hash()` and `X509_issuer_name_hash()`.

Reply:
Thank you @DDvO for your effort. I am sorry for being so demanding. I'll start the final review pass right away.